<a href="https://colab.research.google.com/github/voronv27/projects-in-ai-and-ml/blob/main/HW5/Homework5_GraduateTask.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Task for 6000 level

##Research Medical Image Segmentation Models

Models used to detec polyps include CNNs-based as well as Transformer/hybrid models. I researched both of these types of models.

###CNN-based models

####U-Net
The (probably) best-known model for polyp detection is U-Net. U-Net is a Convolutional Neural Network (CNN) which was able to train on less samples more efficiently via data augmentation.

The architecture of U-Net uses both a contracting path to get context, and a symmetric expanding path for accurate polyp localization (encoder-decoder architecture). In the successive layers of the contracting path, the pooling operators are replaced by upsampling operators which can increase the output resolution. High-resolution features from the contracting path are combined with the supplementary upsampled output in order to localize, and so the convolution layer learns to be more precise in its output. As the contracting and expansive paths are about symmetric, the architecture is U-shaped, hence the name U-Net.

At the time that "[U-Net: Convolutional Networks for Biomedical
Image Segmentation](https://arxiv.org/pdf/1505.04597)" was published, U-Net was able to outperform the previously best method (a sliding-window CNN) and won the ISBI cell tracking challenge in 2015. As reported in the paper, U-Net is also quite fast on GPU, able to segment a 512x512 image in less than a second.

#####Evaluation
U-Net was evaluated on a few segmentation tasks in the paper. The first was the segmentation of neuronal structures in electron microscopic readings, for which U-Net achieved a warping error of 0.0003529, a rand-error of 0.0382, and a pixel error of 0.0611, outperforming the sliding-window convolutional network's best submission. Another task was a cell segmentation task which was part of the ISBI cell tracking challenge 2014 and 2015. In the 2015 challenge, U-NET achieved an average IOU (intersection over union) of 92% on dataset PhC-U373.

#####Common Metrics
In this paper, Dice Similarity Coefficient (DSC) (a popular evaluation metric for medical image segmentation) was not evaluated. IOU, or the Jaccard Index was reported to be 92% on PhC-U373.

#####Sources
[U-Net: Convolutional Networks for Biomedical Image Segmentation](https://arxiv.org/pdf/1505.04597)

####HarDNet-CPS

HarDNet-CPS is a model specifically built for/trained for polyp detection (as opposed to general medical image segmentation). In [HarDNet-CPS: Colorectal polyp segmentation based on Harmonic Densely United Network](https://www.sciencedirect.com/science/article/pii/S1746809423003865), the authors proposed the Harmonic Densely United Network for Colorectal Polyp Segmentation, or HarDNet-CPS for this task.

HarDNet-CPS uses Harmonic-Connected Network-68 as a base. Like U-Net, it has an encoder-decoder structure and is a CNN. However, the structure of this model differs in that the encoder architecture is based on multiple different types of Harmonic Densely Connected Networks (HarDNet). The decoder architecture uses the multi-scale channel attention module (MS-CAM), spacial feature fusion (ASFF) and image data aggregation.

#####Evaluation
The model test set used five open polyp segmentation datasets: Kvasir, CVC-300, CVC-ClinicDB, CVC-ColonDB, and ETIL-LaribPolypDB.

#####Common Metrics
For the test set, HarDNet-CPS had a mean Dice (DSC) of 0.891, and a mean IoU of 0.826. On the same set, U-NET has DSC of 0.710 and IoU 0.627, so HarDNet significantly outperformed U-Net on this test set.

#####Sources
[HarDNet-CPS: Colorectal polyp segmentation based on Harmonic Densely United Network](https://www.sciencedirect.com/science/article/pii/S1746809423003865)

###Transformer / Hybrid Models

####TransUNet

U-Net and other CNN-based models have the typical CNN limitation in modeling long-range dependency, which reduced the model's global perception. As such, transformer models which excel at sequence-to-sequence prediction, became a competing architecture to CNN-based models. However, transformers on their own can have limited localization abilities due to insufficient low-level details, struggling with creating precise boundaries. To try and get the "best of both worlds", TransUNet, a hybrid model using both Transformers and U-Net, was proposed.

In TransUNet, the Transformer is used to encode tokenized image patches from a CNN feature map, allowing for the extraction of global contexts (a weakness of a pure CNN model). Then, the decoder upsamples the encoded features and combines them with the high-resolution CNN feature maps for precise localization (improving upon the weak points of a pure Transformer model). The combined architecture allows the model to use the strong encoding capabilities of a Transformer while enhancing finer details via U-Net (a CNN).

#####Evaluation
In the paper "[TransUNet: Transformers Make Strong
Encoders for Medical Image Segmentation](https://arxiv.org/pdf/2102.04306)", the model was tested against  the Synapse multi-organ CT dataset and ACDC (Automated Cardiac Diagnosis Challenge) dataset.

#####Common Metrics
This paper used the average dice score to compare models. On the Synapse multi-organ CT dataset, the average DSC for TransUNet was 77.48, while U-Net had a score of 74.68. The results showed that TransUNet outperformed other CNN-based and pure Transformer-based encoders. On the ACDC dataset, TransUNet had an average DSC of 89.71.

#####Sources
[TransUNet: Transformers Make Strong
Encoders for Medical Image Segmentation](https://arxiv.org/pdf/2102.04306)

###Polyp-PVT

Similar to TransUNet, Polyp-PVT is another hybrid model, adopting a pyramid vision transformer (PVT) encoder to learn better representations than pure CNN models. It also uses modules CFM, CIM, and SAM in its decoder to improve upon getting polyp information in both high-level (CFM) and low-level (CIM) features, and then fusing them together (SAM).

The results shown in "[Polyp-PVT: Polyp Segmentation with Pyramid
Vision Transformers](https://arxiv.org/pdf/2108.06932)" showed that this model is more robust in challenging situations such as rotation than existing models and methods.

#####Evaluation
This model was evaluated on five public datasets, including KvasirSEG, ClinicDB, ColonDB, Endoscene, and
ETIS. Its performance was compared against various open-source models such as U-Net.

#####Common Metrics
On the Kvasir-SEG dataset, the max Dice (DSC) score was 0.917 for Polyp-PVT, whereas U-NET had a max Dice score of 0.818. The max IoU score for Polyp-PVT was 0.864, compared to U-Net's 0.746. On the ClinicDB dataset, Polyp-PVT had a max DSC of 0.937 and max IoU of 0.889, compared to U-Net's 0.823 and 0.755.

#####Sources
[Polyp-PVT: Polyp Segmentation with Pyramid
Vision Transformers](https://arxiv.org/pdf/2108.06932)

##Implement U-Net

For this task we will be using the following [dataset](https://polyp.grand-challenge.org/CVCClinicDB/).

It has the following attributions:

####Copyright

Images from folder ‘Original’ are property of Hospital Clinic, Barcelona, Spain

Images from folder ‘Ground Truth’ are propery of Computer Vision Center, Barcelona, Spain

####Referencing

The use of this database is completely restricted for research and educational purposes. The use of this database is forbidden for commercial purposes.

If you use this database for your experiments please include the following reference:

Bernal, J., Sánchez, F. J., Fernández-Esparrach, G., Gil, D., Rodríguez, C., & Vilariño, F. (2015). WM-DOVA maps for accurate polyp highlighting in colonoscopy: Validation vs. saliency maps from physicians. Computerized Medical Imaging and Graphics, 43, 99-111 .

We will be getting a train/test split of this database from [huggingface](https://huggingface.co/datasets/Angelou0516/CVC-ClinicDB).

Hugging Face Citations:  
If you use this dataset, please cite:

```
@article{bernal2015wm,
  title={WM-DOVA maps for accurate polyp highlighting in colonoscopy: Validation vs. saliency maps from physicians},
  author={Bernal, Jorge and S{'a}nchez, F Javier and Fern{'a}ndez-Esparrach, Gloria and Gil, Debora and Rodr{\'\i}guez, Cristina and Vilari{\~n}o, Fernando},
  journal={Computerized Medical Imaging and Graphics},
  volume={43},
  pages={99--111},
  year={2015},
  publisher={Elsevier}
}
```

Dataset Split: This train/validation/test split is created by the following study. Please find more details in:

```
@article{chang2024esfpnet,
  title={ESFPNet: Efficient Stage-Wise Feature Pyramid on Mix Transformer for Deep Learning-Based Cancer Analysis in Endoscopic Video},
  author={Chang, Qi and Ahmad, Danish and Toth, Jennifer and Bascom, Rebecca and Higgins, William E},
  journal={Journal of Imaging},
  volume={10},
  number={8},
  pages={191},
  year={2024},
  publisher={MDPI}
}
```

In [32]:
from datasets import load_dataset

# Load dataset
dataset = load_dataset("Angelou0516/CVC-ClinicDB")

Resolving data files:   0%|          | 0/981 [00:00<?, ?it/s]

Resolving data files:   0%|          | 0/123 [00:00<?, ?it/s]

Resolving data files:   0%|          | 0/123 [00:00<?, ?it/s]

In [33]:
# fixes an issue with torchvision imports
import sys
!{sys.executable} -m pip uninstall -y sympy
!{sys.executable} -m pip install sympy==1.13.3

print("sympy==1.13.3 installed successfully.")

Found existing installation: sympy 1.13.3
Uninstalling sympy-1.13.3:
  Successfully uninstalled sympy-1.13.3
  Using cached sympy-1.13.3-py3-none-any.whl.metadata (12 kB)
Using cached sympy-1.13.3-py3-none-any.whl (6.2 MB)
sympy==1.13.3 installed successfully.


In [34]:
import torch
from torchvision import transforms

transform = transforms.Compose([
  transforms.Resize((256, 256)),
  transforms.RandomHorizontalFlip(),
  transforms.RandomRotation(15),
  transforms.ToTensor(),
])

def preprocess(example):
  image = transform(example["image"].convert("RGB"))
  mask = transform(example["mask"].convert("L"))
  mask = (mask > 0).float()

  return {"image": image, "mask": mask}

dataset = dataset.map(preprocess)
dataset.set_format(type="torch", columns=["image", "mask"])

Map:   0%|          | 0/490 [00:00<?, ? examples/s]

Map:   0%|          | 0/61 [00:00<?, ? examples/s]

Map:   0%|          | 0/61 [00:00<?, ? examples/s]

In [35]:
import torch.nn as nn

class DoubleConv(nn.Module):
  def __init__(self, in_ch, out_ch):
    super().__init__()
    self.net = nn.Sequential(
        nn.Conv2d(in_ch, out_ch, 3, padding=1),
        nn.ReLU(inplace=True),
        nn.Conv2d(out_ch, out_ch, 3, padding=1),
        nn.ReLU(inplace=True)
    )

  def forward(self, x):
    return self.net(x)

class UNet(nn.Module):
  def __init__(self):
    super().__init__()

    # encoder (4 layers, first 3 are two conv layers +
    # max pool, last layer skips the max pool)
    self.e1 = DoubleConv(3, 64)
    self.pool1 = nn.MaxPool2d(2)

    self.e2 = DoubleConv(64, 128)
    self.pool2 = nn.MaxPool2d(2)

    self.e3 = DoubleConv(128, 256)
    self.pool3 = nn.MaxPool2d(2)

    self.bottleneck = DoubleConv(256, 512)

    # decoder (3 layers + output)
    self.d3 = nn.ConvTranspose2d(512, 256, 2, 2)
    self.conv3 = DoubleConv(512, 256)

    self.d2 = nn.ConvTranspose2d(256, 128, 2, 2)
    self.conv2 = DoubleConv(256, 128)

    self.d1 = nn.ConvTranspose2d(128, 64, 2, 2)
    self.conv1 = DoubleConv(128, 64)

    # output layer
    self.out = nn.Conv2d(64, 1, 1)

  def forward(self, x):
    # encoder
    down1 = self.e1(x)
    down2 = self.e2(self.pool1(down1))
    down3 = self.e3(self.pool2(down2))
    b = self.bottleneck(self.pool3(down3))

    # decoder
    up3 = self.d3(b)
    up3 = torch.cat([up3, down3], 1)
    up3 = self.conv3(up3)

    up2 = self.d2(up3)
    up2 = torch.cat([up2, down2], 1)
    up2 = self.conv2(up2)

    up1 = self.d1(up2)
    up1 = torch.cat([up1, down1], 1)
    up1 = self.conv1(up1)

    return torch.sigmoid(self.out(up1))

In [36]:
# metrics
def dice_score(pred, target):
  pred = (pred > 0.5).float()
  intersection = (pred * target).sum()
  union = pred.sum() + target.sum()
  return (2 * intersection) / (union + 1e-6)

def iou_score(pred, target):
  pred = (pred > 0.5).float()
  intersection = (pred * target).sum()
  union = pred.sum() + target.sum() - intersection

  return intersection / (union + 1e-6)

In [40]:
def dice_loss(pred, target):
  intersection = (pred * target).sum()
  union = pred.sum() + target.sum()
  dice = (2. * intersection + 1e-6) / (union + 1e-6)
  return 1 - dice

In [47]:
from torch.utils.data import DataLoader

train_loader = DataLoader(dataset["train"], batch_size=8, shuffle=True)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = UNet()
model.to(device)
optimizer = torch.optim.Adam(model.parameters(), lr=0.00001)

total_loss = 0
for epoch in range(30):
  model.train()

  total_loss = 0

  for batch in train_loader:
    images = batch["image"].cuda()
    masks = batch["mask"].cuda()

    preds = model(images)
    loss = dice_loss(preds, masks)

    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    total_loss += loss.item()

  if (epoch % 5) == 0:
    print(f"Epoch {epoch}: Train Loss = {total_loss/len(train_loader):.4f}")

Epoch 0: Train Loss = 0.8383
Epoch 5: Train Loss = 0.7192
Epoch 10: Train Loss = 0.6959
Epoch 15: Train Loss = 0.6904
Epoch 20: Train Loss = 0.6789
Epoch 25: Train Loss = 0.6744


In [49]:
test_loader = DataLoader(dataset["test"], batch_size=1, shuffle=False)
model.eval()
dice_total = 0
iou_total = 0

with torch.no_grad():
  for batch in test_loader:
    images = batch["image"].cuda()
    masks = batch["mask"].cuda()

    preds = model(images)

    dice_total += dice_score(preds, masks).item()
    iou_total += iou_score(preds, masks).item()

print(f"Dice Score: {dice_total/len(test_loader):.4f}")
print(f"IoU Score: {iou_total/len(test_loader):.4f}")

Dice Score: 0.3280
IoU Score: 0.2193


###My Metrics
My simple implementation of U-Net had a DSC of ~33% and IoU of ~22%.

###Model Comparison
In the [Polyp-PVT research paper](https://arxiv.org/pdf/2108.06932), they trained and tested on ClinicDB as well and reported their DSC and IoU, as well as the DSC and IoU of various other models.

 Polyp-PVT had a max DSC of 0.937 and max IoU of 0.889. U-Net had a max DSC of 0.823 and max IoU of 0.755.

 As to be expected, these values are higher than my model, which has a very simple implementation (only 4 layers, basic U-Net instead of U-Net++, no dropout). My model also had less data to train on, since the training dataset in this study consisted of five datasets combined, while I only used ClinicDB. They also had significantly more epochs (100 vs 30), while I kept my number of epochs lower for runtime reasons.

 For these reasons, the models in the Polyp-PVT research paper performed much better than mine on ClinicDB. This comparison serves as a good way to understand the impact of training time, data size, and model complexity on model performance, as even with the same underlying architecture, the U-Net model outperformed mine by quite a lot.